lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -0,0 +1,465 @@
+    return;
+
+  Type const *SrcTy = Cast.getSrcTy();
+  Type *DstVecTy = Cast.getDestTy();
+  // Leave scalar casts as is. Skip boolean vector casts becase there
+  // are no suitable OCL built-ins.
+  if (!DstVecTy->isVectorTy() || SrcTy->getScalarSizeInBits() == 1 ||
+    DstVecTy->getScalarSizeInBits() == 1)
+    return;
+
+  // Assemble built-in name -> convert_gentypeN
+  std::string CastBuiltInName(kOCLBuiltinName::ConvertPrefix);
+  // Check if this is 'floating point -> unsigned integer' cast
+  CastBuiltInName += mapLLVMTypeToOCLType(DstVecTy, !isa<FPToUIInst>(Cast));
+
+  // Replace LLVM conversion instruction with call to conversion built-in
+  BuiltinFuncMangleInfo Mangle;
+  // It does matter if the source is unsigned integer or not. SExt is for
+  // signed source, ZExt and UIToFPInst are for unsigned source.
+  if (isa<ZExtInst>(Cast) || isa<UIToFPInst>(Cast))
+    Mangle.addUnsignedArg(0);
+
+  AttributeList Attributes;
+  CallInst *Call =
+    addCallInst(M, CastBuiltInName, DstVecTy, Cast.getOperand(0), &Attributes,
+      &Cast, &Mangle, Cast.getName(), false);
+  Cast.replaceAllUsesWith(Call);
+  Cast.eraseFromParent();
+}
+
+void SPIRVToOCL::visitCallSPRIVImageQuerySize(CallInst *CI) {
+  Function *Func = CI->getCalledFunction();
+  // Get image type
+  Type *ArgTy = Func->getFunctionType()->getParamType(0);
+  assert(ArgTy->isPointerTy() &&
+    "argument must be a pointer to opaque structure");
+  StructType *ImgTy = cast<StructType>(ArgTy->getPointerElementType());
+  assert(ImgTy->isOpaque() && "image type must be an opaque structure");
+  StringRef ImgTyName = ImgTy->getName();
+  assert(ImgTyName.startswith("opencl.image") && "not an OCL image type");
+
+  unsigned ImgDim = 0;
+  bool ImgArray = false;
+
+  if (ImgTyName.startswith("opencl.image1d")) {
+    ImgDim = 1;
+  }
+  else if (ImgTyName.startswith("opencl.image2d")) {
+    ImgDim = 2;
+  }
+  else if (ImgTyName.startswith("opencl.image3d")) {
+    ImgDim = 3;
+  }
+  assert(ImgDim != 0 && "unexpected image dimensionality");
+
+  if (ImgTyName.count("_array_") != 0) {
+    ImgArray = true;
+  }
+
+  AttributeList Attributes = CI->getCalledFunction()->getAttributes();
+  BuiltinFuncMangleInfo Mangle;
+  Type *Int32Ty = Type::getInt32Ty(*Ctx);
+  Instruction *GetImageSize = nullptr;
+
+  if (ImgDim == 1) {
+    // OpImageQuerySize from non-arrayed 1d image is always translated
+    // into get_image_width returning scalar argument
+    GetImageSize = addCallInst(M, kOCLBuiltinName::GetImageWidth, Int32Ty,
+      CI->getArgOperand(0), &Attributes, CI, &Mangle,
+      CI->getName(), false);
+    // The width of integer type returning by OpImageQuerySize[Lod] may
+    // differ from i32
+    if (CI->getType()->getScalarType() != Int32Ty) {
+      GetImageSize = CastInst::CreateIntegerCast(GetImageSize,
+        CI->getType()->getScalarType(),
+        false, CI->getName(), CI);
+    }
+  }
+  else {
+    assert((ImgDim == 2 || ImgDim == 3) && "invalid image type");
```

Comment:
I don't see any value in this assertion. `ImgDim` might be zero, but the fact that it is not zero has been asserted already